code/android-core-library/src/main/java/com/adobe/marketing/mobile/internal/configuration/ConfigurationExtension.kt

```diff
@@ -0,0 +1,575 @@
+        configurationRulesManager: ConfigurationRulesManager
+    ) : super(extensionApi) {
+        this.serviceProvider = serviceProvider
+        this.appIdManager = appIdManager
+        this.cacheFileService = cacheFileService
+        this.launchRulesEvaluator = launchRulesEvaluator
+        this.retryWorker = retryWorker
+        this.configurationStateManager = configurationStateManager
+        this.configurationRulesManager = configurationRulesManager
+
+        loadInitialConfiguration()
+
+        EventHub.shared.registerEventPreprocessor {
+            launchRulesEvaluator.process(it)
+        }
+    }
+
+    /**
+     * Loads the initial configuration.
+     */
+    private fun loadInitialConfiguration() {
+        val appId = this.appIdManager.loadAppId()
+
+        if (!appId.isNullOrBlank()) {
+            val eventData: MutableMap<String, Any?> =
+                mutableMapOf(
+                    Configuration.CONFIGURATION_REQUEST_CONTENT_JSON_APP_ID to appId,
+                    CONFIGURATION_REQUEST_CONTENT_IS_INTERNAL_EVENT to true
+                )
+            dispatchConfigurationRequest(eventData)
+        }
+
+        val initialConfig: Map<String, Any?> = this.configurationStateManager.loadInitialConfig()
+
+        if (initialConfig.isNotEmpty()) {
+            applyConfigurationChanges(null, RulesSource.CACHE, false)
+        } else {
+            MobileCore.log(
+                LoggingMode.VERBOSE,
+                TAG,
+                "Initial configuration loaded is empty."
+            )
+        }
+    }
+
+    /**
+     * Sets up the Configuration extension for the SDK by registering
+     * event listeners for processing events.
+     */
+    override fun onRegistered() {
+        super.onRegistered()
+
+        // States should not be created in the constructor of the Extension.
+        // Publish initial state that is pre-computed in the constructor when the
+        // registration completes.
+        val initialConfigState = configurationStateManager.environmentAwareConfiguration
+        if (initialConfigState.isNotEmpty()) {
+            publishConfigurationState(initialConfigState, null)
+        }
+
+        api.registerEventListener(
+            EventType.CONFIGURATION,
+            EventSource.REQUEST_CONTENT
+        ) {
+            handleConfigurationRequestEvent(it)
+        }
+    }
+
+    override fun getName(): String {
+        return EXTENSION_NAME
+    }
+
+    override fun getFriendlyName(): String {
+        return EXTENSION_FRIENDLY_NAME
+    }
+
+    override fun getVersion(): String {
+        return EXTENSION_VERSION
+    }
+
+    /**
+     * Responsible for handling the incoming event requests directed towards [ConfigurationExtension].
+     *
+     * @param event the event request dispatched to the [ConfigurationExtension]
+     */
+    internal fun handleConfigurationRequestEvent(event: Event) {
+        if (event.eventData == null) return
```

Comment:
Can we create pending shared state before processing events which change configuration? Each extension processes events in its own queue and this will guarantee they wait for configuration changes.